### PR TITLE
Adds Support for Swift Package Manager

### PR DESCRIPTION
**Pull request by jaerod95**
Adds support for SPM. See README.md for instructions on how to add it to your project.

Finally got around to fixing this up again. Now that Apple supports Resources I was able to import the Localizations. I added to the LTHPasscodeViewControllerStrings method to search for the SPM bundle if it didn't find a localized string since Apple moves those localizations into its own bundle.

### CODE REVIEW

**Review by rolandleth — CHANGES_REQUESTED** [review]
Thank you for the contribution! Just a few changes to be made, please.

Also, could you please explain a bit the need for the `LTHPasscodeViewControllerStrings` change? Haven't worked with SPM yet, so no idea what the issue was and how it got solved.

File: LTHPasscodeViewController/LTHPasscodeViewController.m [hunk 1]
```diff
@@ -16,7 +16,7 @@
 
 #ifndef LTHPasscodeViewControllerStrings
 #define LTHPasscodeViewControllerStrings(key) \
-[[NSBundle bundleWithPath:[[NSBundle bundleForClass:[LTHPasscodeViewController class]] pathForResource:@"LTHPasscodeViewController" ofType:@"bundle"]] localizedStringForKey:(key) value:@"" table:_localizationTableName]
+[[[NSBundle bundleWithPath:[[NSBundle bundleForClass:[LTHPasscodeViewController class]] pathForResource:@"LTHPasscodeViewController" ofType:@"bundle"]] localizedStringForKey:(key) value:@"" table:_localizationTableName] length] == 0 ? NSLocalizedString(key, bundle: .module, comment: "") : [[NSBundle bundleWithPath:[[NSBundle bundleForClass:[LTHPasscodeViewController class]] pathForResource:@"LTHPasscodeViewController" ofType:@"bundle"]] localizedStringForKey:(key) value:@"" table:_localizationTableName]
```
[the remark this answers is not in the document]
Comment (jaerod95) [reply LTHPasscodeViewController/LTHPasscodeViewController.m:19]: Love this idea. To be honest my Objc experience is basically none, so I was kinda just shooting in the dark here to make it work 😅 I'll make the change.

File: Package.swift [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+        .library(
+            name: "LTHPasscodeViewController",
+            targets: ["LTHPasscodeViewController"]
+        ),
+    ],
+    targets: [
+        .target(
+            name: "LTHPasscodeViewController",
+            path: ".",
+            exclude: ["Demo", "CHANGELOG.md", "LICENSE.txt", "README.md"],
```
[the remark this answers is not in the document]
Comment (jaerod95) [reply Package.swift:20]: This doesn't stop it from being shown when it's used in the side bar, but just doesn't compile it into the package binary. If that's still a problem I can process it into the bundle I believe.
Screen Shot 2021-03-29 at 8 08 00 AM

[the remark this answers is not in the document]
Comment (rolandleth) [reply Package.swift:20]: Hmm, I usually include license files in apps (even create About pages to display the full license text), so if it's possible, without breaking anything, that would be nice 👍

File: README.md [hunk 3]
```diff
@@ -1,8 +1,31 @@
...
+
+### CocoaPods
+
+[CocoaPods](https://cocoapods.org) is a dependency manager for Cocoa projects. For usage and installation instructions, visit their website. To integrate LTHPasscodeViewController into your Xcode project using CocoaPods, specify it in your `Podfile`:
+
+```ruby
+pod 'LTHPasscodeViewController', '~> 4.0.1'
+```
+
+### Swift Package Manager
```
[the remark this answers is not in the document]
Comment (jaerod95) [reply README.md:6]: Done

File: README.md [hunk 4]
```diff
@@ -1,8 +1,31 @@
 # LTHPasscodeViewController
 Simple to use iOS 7 style Passcode - the one you get in Settings when changing your passcode.
 
+# Installation
+
+### CocoaPods
+
+[CocoaPods](https://cocoapods.org) is a dependency manager for Cocoa projects. For usage and installation instructions, visit their website. To integrate LTHPasscodeViewController into your Xcode project using CocoaPods, specify it in your `Podfile`:
```
[the remark this answers is not in the document]
Comment (jaerod95) [reply README.md]: Done

File: README.md [hunk 5]
```diff
@@ -1,8 +1,31 @@
...
+```ruby
+pod 'LTHPasscodeViewController', '~> 4.0.1'
+```
+
+### Swift Package Manager
+
+__NOTE__: _These instructions are intended for usage on Xcode 11 and higher. Xcode 11 is the first version of Xcode that integrates Swift Package manager and makes it way easier to use than it was at the command line. If you are using older versions of Xcode, we recommend using CocoaPods._
+
+1. Go to File > Swift Packages > Add Package Dependency...
+2. Paste the URL to the LTHPasscodeViewController repo on GitHub (https://github.com/rolandleth/LTHPasscodeViewController.git) into the search bar, then hit the Next button:
```
[the remark this answers is not in the document]
Comment (jaerod95) [reply README.md]: Done

File: Package.swift (hunk 2 above)
[the remark this answers is not in the document]
Comment (jaerod95) [reply Package.swift:20]: Sounds great I'll make the change.

File: Package.swift [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+            sources: ["LTHPasscodeViewController/LTHKeychainUtils.h", "LTHPasscodeViewController/LTHKeychainUtils.m", "LTHPasscodeViewController/LTHPasscodeViewController.h", "LTHPasscodeViewController/LTHPasscodeViewController.m"],
```
[the remark this answers is not in the document]
Comment (jaerod95) [reply Package.swift]: This actually wasn't needed so I just removed the line.

File: LTHPasscodeViewController/LTHPasscodeViewController.m (hunk 1 above)
[the remark this answers is not in the document]
Comment (jaerod95) [reply LTHPasscodeViewController/LTHPasscodeViewController.m]: Updated to use your snippet. Worked great!

File: README.md [hunk 7]
```diff
@@ -46,7 +69,7 @@ if ([LTHPasscodeViewController doesPasscodeExist]) {
 
 ```objc
 /**
- @param	viewController The view controller where the passcode view controller will be displayed.
+ @param    viewController The view controller where the passcode view controller will be displayed.
```
[the remark this answers is not in the document]
Comment (jaerod95) [reply README.md]: Done

**Comment by jaerod95** [comment]
The reason I needed to change the LTHPasscodeViewControllerStrings was due to how the new SPM tooling processes resources. Basically it compiles them into a bundle and exposes that bundle to Swift and Objc. I gathered most of my information on what that looks like from this article: https://darjeelingsteve.com/articles/How-to-Use-Module-Resources-in-Objective-C-SPM-Packages.html

**Comment by rolandleth** [comment]
Thanks for sharing the article! 

Hmm, is there any way to test the package before merging? Can I add it to a project by entering the URL of your `dev` branch? Not sure what the syntax would be, maybe https://github.com/SimpleNexus/LTHPasscodeViewController/tree/dev?

File: Package.swift (hunk 2 above)
[the remark this answers is not in the document]
Comment (rolandleth) [reply Package.swift:20]: I think this is the only thing remaining, right?

**Comment by jaerod95** [comment]
> Thanks for sharing the article!
> 
> Hmm, is there any way to test the package before merging? Can I add it to a project by entering the URL of your `dev` branch? Not sure what the syntax would be, maybe https://github.com/SimpleNexus/LTHPasscodeViewController/tree/dev?

Yes! You can simply add https://github.com/SimpleNexus/LTHPasscodeViewController as the git source, then on the rules page, just select branch and point it at dev as seen in the picture below:
Screen Shot 2021-04-08 at 10 16 45 AM

File: Package.swift (hunk 2 above)
[the remark this answers is not in the document]
Comment (jaerod95) [reply Package.swift]: Yup, just updated the package.swift to now process the `LICENSE.txt` file.

**Comment by rolandleth** [comment]
Cool, I'll give it a go today, thanks!

**Comment by rolandleth** [comment]
Hey! Very sorry for the long delay, but I managed to give it a go today. I added the package to a Swift project and everything works, except localization. Maybe I'm not doing it properly, but I set the project to use English and German, I added a `Localizable.strings` file, but the passcode view isn't localized.

Either I'm doing something wrong, or the macro isn't working properly 🤔🤔 Can you get localization to work when using the passcode as a package?

The demo project is properly localized.

**Comment by jaerod95** [comment]
@rolandleth Sorry for the delay, got distracted for a quick....8 months haha. However I updated the code to work with localization (nice catch by the way). There was an issue where I was looking in the wrong bundle The last two commits fix that. Let me know if there are any other issues.

**Review by rolandleth — APPROVED** [review]

**Comment by rolandleth** [comment]
@jaerod95 This has to be the slowest moving PR in history 🙈 I tested it and it works, I think it's good to merge, thank you!

**Pull request merged**